### ISSUE TITLE
error when "go get github.com/TBD54566975/ssi-sdk/did"

### ISSUE THREAD

**Issue by enricobottazzi** [TBD54566975/ssi-sdk#104]
```bash 
enricobottazzi@Air-di-Enrico web5 % go get github.com/TBD54566975/ssi-sdk/did
# github.com/TBD54566975/ssi-sdk/cryptosuite
../../../go/pkg/mod/github.com/!t!b!d54566975/ssi-sdk@v0.0.0-20220614001841-7dd06673dbf3/cryptosuite/cryptosuite.go:49:9: undefined: LDKeyType
../../../go/pkg/mod/github.com/!t!b!d54566975/ssi-sdk@v0.0.0-20220614001841-7dd06673dbf3/cryptosuite/jwssignaturesuite.go:25:45: undefined: LDKeyType
../../../go/pkg/mod/github.com/!t!b!d54566975/ssi-sdk@v0.0.0-20220614001841-7dd06673dbf3/cryptosuite/jwssignaturesuite.go:47:35: undefined: LDKeyType```

**Comment by andorsk** [comment]
I've replicated it on my side: 
```➜  did git:(main) ✗ go run main.go
# github.com/TBD54566975/ssi-sdk/cryptosuite
../../cryptosuite/cryptosuite.go:49:9: undefined: LDKeyType
../../cryptosuite/jwssignaturesuite.go:25:45: undefined: LDKeyType
../../cryptosuite/jwssignaturesuite.go:47:35: undefined: LDKeyType
➜  did git:(main) ✗```

**Comment by andorsk** [comment]
I've set up a PR with a temporary fix: https://github.com/TBD54566975/ssi-sdk/pull/109. 

Off LDKeyTypeFix branch:
```➜  did git:(LDKeyTypeFix) ✗ go run main.go
Works%
```

Off main:
```
➜  did git:(main) ✗ go run main.go
# github.com/TBD54566975/ssi-sdk/cryptosuite
../../cryptosuite/cryptosuite.go:49:9: undefined: LDKeyType
../../cryptosuite/jwssignaturesuite.go:25:45: undefined: LDKeyType
../../cryptosuite/jwssignaturesuite.go:47:35: undefined: LDKeyType
➜  did git:(main) ✗```

Oddly, mage test passes, which includes very similar code. 

Room to improve this by replicating undefined issue over a test. 

```
package main

import (

	"github.com/TBD54566975/ssi-sdk/did"
	"github.com/TBD54566975/ssi-sdk/crypto"

)

func main() {
	// Generate a DID key.
	pk, _, err := crypto.GenerateEd25519Key()
	if err != nil {
		panic(err)
	}

	_, err = did.CreateDIDKey(crypto.Ed25519, pk)
	if err != nil {
		panic(err)
	}
	print("Works")
}```

**Comment by decentralgabe** [comment]
The issue is relating to using the [jwx library](https://github.com/lestrrat-go/jwx) which requires that go build tag to use the secp256k1 curve key.

The tag does not need to be provided in the file, but must be provided as a argument to the `go` command, which is what the mage file does here -- https://github.com/TBD54566975/ssi-sdk/blob/main/magefile.go#L30

**Comment by andorsk** [comment]
I see. Thanks for the merge! It's my first contribution to the repo. 

What's the suggested long term way to handle this? Currently, I am able to compile a sample without adding the `secp256k1` tag in the build:

```
➜  did git:(examples) ✗ go build; ls -lh did
-rwxr-xr-x  1 akmb2  staff   8.4M Jun 25 14:36 did
```
this is with PR 109 in place. I'm not sure if this has downstream affects. 

For someone that just wants to use the `ssi-sdk`, it seems like asking them to remember to add build options might be considered unreasonable. 

For example: Let's say I'm using the SSI-SDK for some other application and I'm just trying to use it out of the box. I shouldn't be forced to know that I need to build with the tag. 

So boiling it down here's two questions I have: 

1. Currently, without the `secp256k1` tag to the build, what would break and how? 
2. What's the long term strategy for tagging builds for projects leveraging using the ssi-sdk?

As an adjacent question: I think one of the things I'm trying to understand conceptually is the scope of the ssi-sdk and what it looks like when it's mature. I'm not sure if there are any documents to help guide that conversation that I haven't seen.

**Comment by andorsk** [comment]
Also, should this issue remain open or can it be closed, with a possibly new issue being made to request any issues with the current implementation?

**Comment by decentralgabe** [comment]
The downstream affect would only be (I think...) when you use a codepath that uses the jwx lib. 

I do agree the build tag is annoying, and I'm not sure an elegant way around it. I looked for other libraries that had good jw* support with secp256k1 and really couldn't find any. I did [open an issue on the jwx lib](https://github.com/lestrrat-go/jwx/issues/670) but didn't get a satisfying answer.

> Currently, without the secp256k1 tag to the build, what would break and how?
JWX usage of the secp256k1 key. Because of the file structure, it's likely that anything using the JWX lib would break. Right now, this is limited to `JsonWebKey2020` and `JsonWebSignature2020` usage, in addition to JWS signing/verification without that suite.

> What's the long term strategy for tagging builds for projects leveraging using the ssi-sdk?

I see a few paths forward:
1. Find a nice workaround to not propagate the build tag to all dependencies
2. Find a different library that has all the same functionality but without the build tag
3. Propose or influence a change in the JWX library
4. Drop secp256k1 key support

(3) and (4) seem the toughest, so (1) and (2) should be explored.
If there are any go experts out there that have additional ideas, of course I welcome them.

> As an adjacent question: I think one of the things I'm trying to understand conceptually is the scope of the ssi-sdk and what it looks like when it's mature. I'm not sure if there are any documents to help guide that conversation that I haven't seen. The README seems to be the best "vision" documentation.

You are right the README is the best of the vision documents. I created [an issue](https://github.com/TBD54566975/ssi-sdk/issues/111) to track this.

**Comment by andorsk** [comment]
I'd suggest there's a 5th (possibly unreasonable option), which is building secp256k1 suport directly into the ssi-sdk rather than rely on a dependency. Looks like a lot of the lib is wrapping the C libs underneath it: 

https://github.com/bitherhq/go-bither/blob/v1.7.0/crypto/secp256k1/secp256.go#L17

I don't have a preference here yet, but I do have a question about regulating our dependency control. What's the criteria for allowing `jwx` into the build and more importantly, in general what guidelines/limits are we going to put on dependencies?

I think this relates loosely to https://github.com/TBD54566975/ssi-sdk/issues/111 so tagging it here. 

Seems like short term with only 7 contributors in general the tagging is an annoying but non-critical issue we could easily document on the README, but in the context of large scale adoption with the community I would think the problem severity escalates. 

https://github.com/tendermint/tendermint/blob/master/crypto/secp256k1/secp256k1.go has one secp256k1 implementation maybe worth a look.

**Comment by decentralgabe** [comment]
I agree native implementation is an option worth considering. The main concern I have is keeping a consistent API for using JW* objects. If we can achieve that without relying on the JWX lib for secp256k1, then let's do it.

**Comment by decentralgabe** [comment]
Hi @enricobottazzi can you still reproduce this issue? I believe it should be fixed after #126

**Comment by enricobottazzi** [comment]
Solved :)
